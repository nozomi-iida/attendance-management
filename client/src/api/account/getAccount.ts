import {HttpClient} from "lib/axios";
import {Account} from "api/account/index";

export async function getAccount(id: number) {
  const res = await HttpClient.get<Account>(`/accounts/${id}`);
  return res.data;
}
