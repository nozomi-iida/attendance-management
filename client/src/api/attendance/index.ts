export type Attendance = {
  id: number;
  breakStartTime: string | null;
  workTime: number;
  breakTime: number;
  startedAt: string;
  endedAt: string | null;
};

export const mockAttendance = (
  modification?: Partial<Attendance>
): Attendance => {
  return {
    id: 1,
    breakStartTime: null,
    workTime: 180,
    breakTime: 30,
    startedAt: new Date().toISOString(),
    endedAt: null,
    ...modification,
  };
};
