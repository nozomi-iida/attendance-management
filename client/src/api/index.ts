export type MutationVariables<T, U> = T extends undefined
  ? {
      urlParams?: T;
      requestBody: U;
    }
  : U extends undefined
  ? {
      urlParams: T;
      requestBody?: U;
    }
  : {
      urlParams: T;
      requestBody: U;
    };

export type QueryVariables<T, U> = T extends undefined
  ? {
      urlParams?: T;
      queryParams: U;
    }
  : U extends undefined
  ? {
      urlParams: T;
      queryParams?: U;
    }
  : {
      urlParams: T;
      queryParams: U;
    };
