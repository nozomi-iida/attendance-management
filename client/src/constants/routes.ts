export const routes = {
  signUp: () => "/sign_up",
  login: () => "/login",
  slackAuth: () => "/slack_auth",
  managements: () => "/",
};
