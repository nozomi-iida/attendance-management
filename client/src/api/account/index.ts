import { Attendance, mockAttendance } from "api/attendance";

export enum AccountRole {
  GENERAL = "general",
  ADMIN = "admin",
}

export type Account = {
  id: number;
  handleName: string;
  email: string;
  role: AccountRole;
  currentAttendance: Attendance | null;
};

export const mockAccount = (modification?: Partial<Account>): Account => {
  return {
    id: 1,
    handleName: "test",
    email: "test@test.com",
    role: AccountRole.GENERAL,
    currentAttendance: mockAttendance(),
    ...modification,
  };
};
