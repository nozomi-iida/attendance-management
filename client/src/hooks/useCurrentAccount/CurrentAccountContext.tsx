import {createContext, FC, ReactNode, useEffect, useState} from "react";
import {Account, mockAccount} from "api/account";
import {PersistKeys} from "constants/persistKeys";
import {getAccount} from "api/account/getAccount";

type CurrentAccountContext = {
  account?: Account;
  setAccount: (account: Account) => void;
};

type CurrentAccountProviderProps = {
  children: ReactNode;
};

export const currentAccountContext = createContext<CurrentAccountContext>({
  account: undefined,
  setAccount: () => undefined,
});

export const CurrentAccountProvider: FC<CurrentAccountProviderProps> = ({
  children,
}) => {
  // TODO
  const [account, setAccount] = useState<Account>(mockAccount);
  useEffect(() => {
    const token = localStorage.getItem(PersistKeys.AuthToken)
    if(!token) return;

    const [, p] = token.split(".");
    const {id} = JSON.parse(window.atob(p));

    if (id) {
      getAccount(id).then((data) => {
        setAccount(data);
      });
    }
  }, []);

  return (
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    <currentAccountContext.Provider value={{ account, setAccount }}>
      {children}
    </currentAccountContext.Provider>
  );
};
