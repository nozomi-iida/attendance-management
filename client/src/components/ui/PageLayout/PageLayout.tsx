import ProLayout, { MenuDataItem } from "@ant-design/pro-layout";
import { FC, ReactNode, useMemo } from "react";
import Logo from "assets/images/logo.png";
import {
  CalendarOutlined,
  ClockCircleOutlined,
  UserOutlined,
} from "@ant-design/icons";
import { routes } from "constants/routes";
import { useCurrentAccount } from "hooks/useCurrentAccount/useCurrentAccount";
import { Button, Popover } from "antd";
import { Link } from "react-router-dom";

type PageLayoutProps = {
  children: ReactNode;
};

const menu: MenuDataItem[] = [
  {
    path: routes.managements(),
    name: "勤怠管理",
    icon: <ClockCircleOutlined />,
  },
  {
    path: routes.accountManagement(),
    name: "社員の勤怠状況",
    icon: <UserOutlined />,
  },
  {
    path: routes.lightningTalks(),
    name: "LTアドベントカレンダー",
    icon: <CalendarOutlined />,
  },
];

export const PageLayout: FC<PageLayoutProps> = ({ children }) => {
  const { account, logout } = useCurrentAccount();
  const rightContentRender = useMemo(
    () =>
      account && (
        <Popover
          placement="bottomRight"
          title="アカウントサービス"
          content={
            <Button type="link" danger onClick={logout}>
              ログアウト
            </Button>
          }
        >
          <Button type="text">{account.handleName}</Button>
        </Popover>
      ),
    [account, logout]
  );

  return (
    <ProLayout
      menuDataRender={() => menu}
      title="SIMULA.Labs"
      logo={Logo}
      rightContentRender={() => rightContentRender}
      fixSiderbar
      // eslint-disable-next-line react/no-unstable-nested-components
      menuItemRender={(item: MenuDataItem, defaultDom: ReactNode) => (
        <Link to={item.path ?? ""}>{defaultDom}</Link>
      )}
    >
      {children}
    </ProLayout>
  );
};
