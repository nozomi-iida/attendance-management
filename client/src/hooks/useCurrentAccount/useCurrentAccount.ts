import { useContext } from "react";
import { PersistKeys } from "constants/persistKeys";
import { getAccount } from "api/account/getAccount";
import { useNavigate } from "react-router-dom";
import { routes } from "constants/routes";
import { CurrentAccountContext } from "./CurrentAccountContext";

export const useCurrentAccount = () => {
  const { account, setAccount } = useContext(CurrentAccountContext);
  const navigate = useNavigate();

  const getCurrentAccount = async (id?: number) => {
    let accountId = id;
    if (!accountId && account) {
      accountId = account.id;
    }
    if (!accountId) return;

    try {
      const res = await getAccount(accountId);
      setAccount(res);
    } catch (e) {
      throw new Error("エラー");
    }
  };

  const logout = () => {
    localStorage.clear();
    setAccount(undefined);
    navigate(routes.login());
  };

  return {
    account,
    getAccount: getCurrentAccount,
    logout,
  };
};
