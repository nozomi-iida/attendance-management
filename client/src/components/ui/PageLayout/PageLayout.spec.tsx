import { render, screen, waitFor } from "@testing-library/react";
import { PageLayout } from "components/ui/PageLayout/PageLayout";
import { createMemoryHistory } from "history";
import { BrowserRouter, Router } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { CurrentAccountContext } from "hooks/useCurrentAccount/CurrentAccountContext";
import { mockAccount } from "api/account";
import { routes } from "constants/routes";

test("should match snapshot", () => {
  const { container } = render(
    <PageLayout>
      <div />
    </PageLayout>,
    { wrapper: BrowserRouter }
  );

  // eslint-disable-next-line testing-library/no-node-access
  expect(container.firstChild).toMatchSnapshot();
});

describe("PageLayout", () => {
  test("logout", async () => {
    const account = mockAccount();
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
        <CurrentAccountContext.Provider
          value={{ account, setAccount: jest.fn }}
        >
          <PageLayout>
            <div />
          </PageLayout>
        </CurrentAccountContext.Provider>
      </Router>
    );
    userEvent.hover(
      screen.getByRole("button", {
        name: account.handleName,
      })
    );
    await waitFor(() => {
      screen.getByText("ログアウト");
    });
    userEvent.click(screen.getByText("ログアウト"));
    await waitFor(() => expect(history.location.pathname).toBe(routes.login()));
  });
});
