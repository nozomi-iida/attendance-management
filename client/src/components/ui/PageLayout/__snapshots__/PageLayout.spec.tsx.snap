// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  class="ant-design-pro ant-pro-basicLayout screen-md ant-pro-basicLayout-side"
>
  <section
    class="ant-layout ant-layout-has-sider"
    style="min-height: 100%;"
  >
    <aside
      class="ant-layout-sider ant-layout-sider-dark ant-pro-sider ant-pro-sider-layout-side"
      style="overflow: hidden; flex: 0 0 208px; max-width: 208px; min-width: 208px; width: 208px;"
    >
      <div
        class="ant-layout-sider-children"
      >
        <div
          class="ant-pro-sider-logo"
          id="logo"
        >
          <a>
            <img
              alt="logo"
              src="logo.png"
            />
            <h1>
              SIMULA.Labs
            </h1>
          </a>
        </div>
        <div
          style="flex: 1; overflow-y: auto; overflow-x: hidden;"
        >
          <ul
            class="ant-menu ant-menu-root ant-menu-inline ant-menu-dark ant-pro-sider-menu"
            data-menu-list="true"
            role="menu"
            style="width: 100%;"
            tabindex="0"
          >
            <li
              class="ant-menu-item ant-menu-item-only-child"
              data-menu-id="rc-menu-uuid-test-/dd93076c8360ebb9cc22e5fe5dcdf7dd624b4fa6c9966963bceaa08f9c5adc0c"
              role="menuitem"
              style="padding-left: 16px;"
              tabindex="-1"
            >
              <span
                class="ant-menu-title-content"
              >
                <span
                  class="ant-pro-menu-item"
                >
                  <span
                    aria-label="clock-circle"
                    class="anticon anticon-clock-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="clock-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                      />
                      <path
                        d="M686.7 638.6L544.1 535.5V288c0-4.4-3.6-8-8-8H488c-4.4 0-8 3.6-8 8v275.4c0 2.6 1.2 5 3.3 6.5l165.4 120.6c3.6 2.6 8.6 1.8 11.2-1.7l28.6-39c2.6-3.7 1.8-8.7-1.8-11.2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="ant-pro-menu-item-title"
                  >
                    勤怠管理
                  </span>
                </span>
              </span>
            </li>
          </ul>
          <div
            aria-hidden="true"
            style="display: none;"
          />
        </div>
        <div
          class="ant-pro-sider-links"
        >
          <ul
            class="ant-menu ant-menu-root ant-menu-inline ant-menu-dark ant-pro-sider-link-menu"
            data-menu-list="true"
            role="menu"
            tabindex="0"
          >
            <li
              class="ant-menu-item ant-menu-item-only-child ant-pro-sider-collapsed-button"
              data-menu-id="rc-menu-uuid-test-collapsed"
              role="menuitem"
              style="padding-left: 16px;"
              tabindex="-1"
            >
              <span
                class="ant-menu-title-content"
              >
                <span
                  aria-label="menu-fold"
                  class="anticon anticon-menu-fold"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="menu-fold"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 000 13.8z"
                    />
                  </svg>
                </span>
              </span>
            </li>
          </ul>
          <div
            aria-hidden="true"
            style="display: none;"
          />
        </div>
      </div>
    </aside>
    <div
      class="ant-layout"
      style="position: relative;"
    >
      <header
        class="ant-layout-header ant-pro-fixed-header-action ant-pro-header-dark"
        style="padding: 0px; height: 48px; line-height: 48px; width: 100%; z-index: 19;"
      >
        <div
          class="ant-pro-global-header ant-pro-global-header-layout-side"
        >
          <div
            style="flex: 1;"
          />
        </div>
      </header>
      <main
        class="ant-layout-content ant-pro-basicLayout-content ant-pro-basicLayout-has-header"
      >
        <div />
      </main>
    </div>
  </section>
</div>
`;
